### PR TITLE
ENH: conditional required annotation for schema

### PR DESCRIPTION
**Pull request by chenqi0805**
### Description
This PR 
* introduces @ConditionalRequired annotation to express conditional required attribute in schema
* modify json schema generator to generate the if-then-else schema based off @ConditionalRequired annotation (Ref: https://github.com/victools/jsonschema-generator/blob/7a5cb07f81f446f149483bcf4b1ba7e2831e6868/jsonschema-examples/src/main/java/com/github/victools/jsonschema/examples/IfThenElseExample.java)
* backfill 
 
### Issues Resolved
Resolves #[Issue number to be closed when this PR is merged]
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by shenkw1 — APPROVED** [review]

File: data-prepper-plugins/date-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/date/DateProcessorConfig.java [hunk 1]
```diff
@@ -20,6 +21,16 @@
 import java.util.Locale;
 import java.time.format.DateTimeFormatter;
 
+@ConditionalRequired(value = {
+        @ConditionalRequired.IfThenElse(
+                ifFulfilled = {@ConditionalRequired.SchemaProperty(field = "match", value = "null")},
```
Comment (san81) [inline data-prepper-plugins/date-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/date/DateProcessorConfig.java:26]: Nit: Having a static import for `@ConditionalRequired.SchemaProperty` probably will make this statement more easy to read

File: data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/AddEntryProcessorConfig.java [hunk 2]
```diff
@@ -14,10 +14,42 @@
...
+                ifFulfilled = {
+                        @ConditionalRequired.SchemaProperty(field = "format", value = "null"),
+                        @ConditionalRequired.SchemaProperty(field = "value", value = "null"),
+                },
+                thenExpect = {@ConditionalRequired.SchemaProperty(field = "value_expression")}
+        ),
+        @ConditionalRequired.IfThenElse(
+                ifFulfilled = {
+                        @ConditionalRequired.SchemaProperty(field = "format", value = "null"),
+                        @ConditionalRequired.SchemaProperty(field = "value_expression", value = "null"),
```
Comment (san81) [inline data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/AddEntryProcessorConfig.java:41]: Is there a way to express `OR` releation when there are more than one condition grouped under `ifFulfilled`

Comment (chenqi0805) [reply data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/AddEntryProcessorConfig.java:41]: For OR relation we have to use separate IfThenElse annotation

Comment (san81) [reply data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/AddEntryProcessorConfig.java]: goit 👍

**Review by san81 — APPROVED** [review]

**Review by oeyh — APPROVED** [review]

**Pull request merged**